### PR TITLE
Encoded Image uploads and retrieval

### LINKED ISSUE #1080: Encoded Image Storage
**Describe the bug**

**To Reproduce**

1.
2.
3.
4.

**Expected behavior**

**Actual behavior**
A clear and concise description of how the code performed w.r.t expectations.

**Screenshots**

**Additional details**

**Potential internship candidates**
Please read this if you are planning to apply for a Palisadoes Foundation internship https://github.com/PalisadoesFoundation/talawa/issues/359

### PR DESCRIPTION
**Pull request by kb-0311**
**What kind of change does this PR introduce?**

bugFix

**Issue Number:**

Fixes #1080

**Did you add tests for your changes?**
Not right now

**Snapshots/Videos:**

**If relevant, did you update the documentation?**

**Summary**

**Does this PR introduce a breaking change?**

**Other information**

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-api/blob/master/CONTRIBUTING.md)?**

Yes

### CODE REVIEW

**Comment by kb-0311** [comment]
pls note that this is an incomplete PR , I will need to manually test these changes with the client side too.

It was getting late where I am from and tomorrow is a weekend so I can focus on complete testing for this one .This PR will be complete tomorrow.

**Comment by kb-0311** [comment]
Update: I got it working. Now images at least can be uploaded by the mobile app and admin portal.

Here is the random screenshot from my pc I uploaded as an image to create the organization as you can see it works.
Screenshot from 2023-02-24 12-54-38
.
The file can be accessed in the express server from the client app like this.
Screenshot from 2023-02-24 12-56-36

I will be submitting a PR to make considerable changes to the admin portal so that this functionality can be used in talawa-admin, tomorrow.
 
@noman2002 I don't know how file reading and image uploads work in flutter (I am not a mobile dev) so I did not test there.

**Comment by palisadoes** [comment]
This looks like it's uploading to the file system. I'm assuming the long term fix would be covered in a GSoC idea.

Remember to add the tests. We need to know if image upload ever breaks again.

**Comment by kb-0311** [comment]
@palisadoes this uploads to the mongoose database first in EncodedImage model , as an encoded base64 url to check for duplicate uploads. the same binary content will not be added to the database rather the fileName pointer will be added to any entity to  which uploads same binary data.

**Comment by kb-0311** [comment]
@palisadoes Hmm I am not so sure about this as a GSoC idea. This was essentially a broken functionality wherein the client-side apps could not upload images as demonstrated in the talawa call. 

Basically what I am saying is this is a hindrance for contributors working on talwa-admin and talawa mobile app. Which in my opinion, would also not allow these contributors to freely end-to-end test the client applications during upcoming GSoC period. 

That is why this content upload thingy should be given a higher priority and should be fixed before GSoC period. So strategies on how to go about this should be well discussed and vetted.

**Review by xoldyckk — APPROVED** [review]
Hmmm. A bit hacky approach but should work.

**Review by xoldyckk — COMMENTED** [review]
@kb-0311 Since clients can't upload images anymore. What image url are they sending to the server?

File: src/utilities/encodedImageStorage/uploadEncodedImage.ts [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+  if (!fs.existsSync(path.join(__dirname, "../../../images"))) {
+    fs.mkdir(path.join(__dirname, "../../../images"), (err) => {
+      if (err) {
+        throw err;
+      }
+    });
+  }
+
+  await writeFile(path.join(__dirname, "../../../" + id), buf);
```
[the remark this answers is not in the document]
Comment (xoldyckk) [reply src/utilities/encodedImageStorage/uploadEncodedImage.ts:50]: Nvm it's included in id

**Comment by xoldyckk** [comment]
@kb-0311 have u tried testing it multiple times instead of just one? Could be that tests start failing after one try cuz of that images folder?

**Comment by kb-0311** [comment]
1. Actually, the image upload was never functional for the talawa client apps. I had asked noman and others whether they had tried it out and they did say no. This PR is a suboptimal solution but I was asked to implement encoding to make sure at least the new contributors working on the client apps at least now can upload and retrieve content. A more optimal strategy will need to be discussed after yoga migration.
2. Yes the client apps will be sending a base64 encoded image for now. I will be submitting a pr to make these changes to the talawa-admin portal so that people over there can now use this function.
Screenshot from 2023-02-24 18-15-23
3. Yeah I have tested it multiple times it probably won't give an error since I am completely avoiding any streams and instead making use of the node's inbuilt 'fs/promises' library to do my tasks. 

Also I should mention that the `uploadEncodedImage.ts` file does not have a 100% codecov because the only lines left uncovered deal with if statement of  the creation of `images/` directory ...  to make sure those lines are covered I will have to delete that directory recusively which I do not want to do for obvious reasons.

**Comment by xoldyckk** [comment]
> @xoldyckk
> 
> 1. Actually, the image upload was never functional for the talawa client apps. I had asked noman and others whether they had tried it out and they did say no. This PR is a suboptimal solution but I was asked to implement encoding to make sure at least the new contributors working on the client apps at least now can upload and retrieve content. A more optimal strategy will need to be discussed after yoga migration.
> 2. Yes the client apps will be sending a base64 encoded image for now. I will be submitting a pr to make these changes to the talawa-admin portal so that people over there can now use this function. This is the change I made to the admin portal code while manual client-server testing.
>    Screenshot from 2023-02-24 18-15-23
> 3. Yeah I have tested it multiple times it probably won't give an error since I am completely avoiding any streams and instead making use of the node's inbuilt 'fs/promises' library to do my tasks.
> 
> Also I should mention that the `uploadEncodedImage.ts` file does not have a 100% codecov because the only lines left uncovered deal with if statement of the creation of `images/` directory ... to make sure those lines are covered I will have to delete that directory recusively which I do not want to do for obvious reasons.

Wouldn't this literally require changing image upload implemenatation on both admin and mobile? If image upload isn't working for createOrganization it shouldn't work for any other resolvers as well right?

**Comment by kb-0311** [comment]
@xoldyckk yep you are precisely right. 

What I am saying is none of the mutations/query in either of the applications mobile app/ admin portal ever sent a file (Upload Scalar) ever as a mutation on the apollo client. The utilities in the API basically existed as a placeholder logic.

I guess you are worried about a breaking change in the frontend but don't worry none of the resolvers in the frontend sends an image file (the $file can be type null or undefined btw ).

**Comment by kb-0311** [comment]
Um why did @Tajcore get removed from reviewers list automatically after requesting review from @xoldyckk ?

**Comment by rishav-jha-mech** [comment]
I have a strong suggestion for this issue,

Can we send the path of the image as `http://localhost:4000/images/{imageName}` instead of `images/{imageName}` which is currently happening. (Because its better than putting the path of the `backendurl/images/{imageName}` everywhere in the image source of `talawa-admin`)

**Comment by kb-0311** [comment]
@rishav-jha-mech No. You should not hardcode the backend URL of a field the backend is not run on the same port for everyone and it will not work on production with a different backend url.

**Comment by rishav-jha-mech** [comment]
> @rishav-jha-mech No. You should not hardcode the backend URL of a field the backend is not run on the same port for everyone and it will not work on production with a different backend url.

Yes that's exactly why i am suggesting to go for the absolute url `backendurl/images/{fileName}` instead of relative url `images/{fileName}`.

**Comment by rishav-jha-mech** [comment]
Change this line in src\utilities\encodedImageStorage\uploadEncodedImage.ts

```
24. if (encodedImageAlreadyExist) {
25.    return process.env.NODE_ENV === "test"
26.      ? `http://localhost:4000/${encodedImageAlreadyExist.fileName}`
27.      : `http://calico.palisadoes.org/talawa/${encodedImageAlreadyExist.fileName}`;
28.  }
```

and

```
53.  return process.env.NODE_ENV === "test"
54.    ? `http://localhost:4000/${uploadedEncodedImage.fileName}`
55.    : `http://calico.palisadoes.org/talawa/${uploadedEncodedImage.fileName}`;
```

The relative path will get changed to absolute one, (you can optimize this as you want). But don't forget to add this in .env file 
```
NODE_ENV=test
```

otherwise the file will be saved as `http://calico.palisadoes.org/talawa/${uploadedEncodedImage.fileName}`

Happy to help 🙌

**Comment by kb-0311** [comment]
@rishav-jha-mech CONFUSED SOUL. :skull:  

The `uploadEncodedImage` is a utility all it does is it sends the file name from the DB to the object in a model in a resolver. the resolver could be signup, createOrg etc which call on this utility and stores the path it returns into their database object as a string. 

I told you that it is a bad idea to hardcode a URL in the frontend ... and you reached the conclusion that somehow it will be a good idea to add absolute values in the backend :skull: ... the problem will be the exact same either way. Not only that we will be storing the absolute value you are suggesting in the database too. Hardcoded backend url... I hope you understand why that is a very bad solution now.

**Comment by rishav-jha-mech** [comment]
@kb-0311 I agree i'm not too familiar with ExpressJs (I have more experience using Django serialisers, where its easy to convert relative urls to absolute urls, Django does it itself). Yeah i was wrong about saving the absolute url to the database 😂, you have made me realise that. What i was thinking, if an image is just a string and mongodb is storing its path, then there must be some way with which this can be changed when sent from the server, i had to test this functionality and i did it in the most horrible way (But it works right ? 😂). Is there a way we can intercept the query and convert image's relative url into absolute url ? or i have to look for some other way on the UI side 😰

**Review by xoldyckk — APPROVED** [review]

**Review by palisadoes — APPROVED** [review]

**Pull request merged**